src/Chapter3.hs

```diff
@@ -479,6 +506,30 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data MentalSpace = Church | Library deriving (Show)
+data House = House Int deriving (Show) -- Int represents number of people inside
```

Comment:
According to our task, only 1, 2, 3 or 4 people can live inside a house. I see that the function `buildHouse` takes a `House` but the only way to construct `House` is to use the `House` constructor. It means that it is possible to create `House (-1)` or `House 15`. You can implement a separate function to restrict the number of people.

This approach is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂